piptools/utils.py

```diff
@@ -260,3 +268,88 @@ def get_hashes_from_ireq(ireq):
     return result
+
+
+def get_compile_command(click_ctx):
+    """
+    Returns a normalized compile command depending on cli context.
+
+    The command will be normalized by:
+        - expanding options short to long
+        - removing values that are already default
+        - sorting the arguments
+        - removing one-off arguments like '--upgrade'
+        - removing arguments that don't change build behaviour like '--verbose'
+    """
+    from piptools.scripts.compile import cli
+
+    # Map of the compile cli options (option name -> click.Option)
+    compile_options = {option.name: option for option in cli.params}
+
+    left_args = []
+    right_args = []
+
+    for option_name, value in click_ctx.params.items():
+        option = compile_options[option_name]
+
+        # Get the latest option name (usually it'll be a long name)
+        option_long_name = option.opts[-1]
+
+        # Collect variadic args separately, they will be added
+        # at the end of the command later
+        if option.nargs < 0:
+            right_args.extend(value)
+            continue
+
+        # Exclude one-off options (--upgrade/--upgrade-package/--rebuild/...)
+        # or options that don't change compile behaviour (--verbose/--dry-run/...)
+        if option_long_name in COMPILE_EXCLUDE_OPTIONS:
+            continue
+
+        # Skip options without a value
+        if option.default is None and not value:
+            continue
+
+        # Skip options with a default value
+        if option.default == value:
+            continue
+
+        # Use a file name for file-like objects
+        if (
+            hasattr(value, "write")
+            and hasattr(value, "read")
+            and hasattr(value, "name")
+        ):
+            value = value.name
+
+        # Convert value to the list
+        values = value
+        if not isinstance(values, (tuple, list)):
+            values = [value]
+
+        for val in values:
+            # Flags don't have a value, thus add to args true or false option lone name
+            if option.is_flag:
+                # If there are false-options, choose an option name depends on a value
+                if option.secondary_opts:
+                    # Get the latest false-option
+                    secondary_option_long_name = option.secondary_opts[-1]
+                    arg = option_long_name if val else secondary_option_long_name
+                # There are no false-options, use true-option
+                else:
+                    arg = option_long_name
+                left_args.append(arg)
+            # Append to args the option with a value
+            else:
+                left_args.append(
+                    "{option}={value}".format(option=option_long_name, value=val)
+                )
+
+    command = "pip-compile"
+
+    if left_args:
+        command += " " + " ".join(sorted(left_args))
+    if right_args:
+        command += " " + " ".join(sorted(right_args))
```

Comment:
" ".join(["pip-compile"] + sorted(left_args) + sorted(right_args))

Reply:
Would [itertools.chain()](https://docs.python.org/3/library/itertools.html#itertools.chain) be useful here to avoid building the temporary list in memory?

Reply:
Barely.

On python 3.6 ish you can use:

`" ".join(["pip-compile", *sorted(left_args), *sorted(right_args)]`

Reply:
Also join without an intermediate list is actually much slower

Reply:
> `" ".join(["pip-compile"] + sorted(left_args) + sorted(right_args))`

I'd prefer first this one, since there is no supposed to be huge amount of args.